spec_util/meld.go

```diff
@@ -271,9 +271,76 @@ func haveCompatibleTypes(dst, src *pb.Primitive) bool {
 		return false
 	}
 
-	cmpDst := &pb.Primitive{Value: dst.Value, FormatKind: dst.FormatKind}
-	cmpSrc := &pb.Primitive{Value: src.Value, FormatKind: src.FormatKind}
-	return proto.Equal(cmpDst, cmpSrc)
+	// First, check that the base types can join.
+	baseJoin := joinBaseTypes(dst, src)
+	if baseJoin == nil {
+		return false
+	}
+
+	// Types are compatible if the base types can join and the format kinds
+	// are equal.
+	cmpDst := &pb.Primitive{Value: baseJoin.Value, FormatKind: dst.FormatKind}
+	cmpSrc := &pb.Primitive{Value: baseJoin.Value, FormatKind: src.FormatKind}
+	if proto.Equal(cmpDst, cmpSrc) {
+		return true
+	}
+
+	// Types are compatible if the base types can join and least one type does
+	// not have any data formats identified.
+	if dst.FormatKind == "" || src.FormatKind == "" {
+		return true
+	}
+
+	return false
+}
+
+// Returns a new Primitive with the Value set as the type-theoretic join of
+// dst.Value and src.Value.  For example, join(int32, uint32) = int64.
+// Returns nil if no such join exists, e.g. join(int64, uint64) = nil.
+func joinBaseTypes(dst, src *pb.Primitive) *pb.Primitive {
```

Comment:
Do we want to also join ints with floating points? Maybe not if the int has a format kind attached, but it may make sense otherwise.

If we choose not to join ints with floating points, the reasoning should at least be documented.

Reply:
I had been thinking about data formats, but now I'm not sure there's a meaningful difference between discarding formats within the same base type vs. from in to float.  I'll extend the join.